extension-manifest-v2/src/classes/BugDb.js

```diff
@@ -10,246 +10,85 @@
-			db.patterns.regex[id] = new RegExp(regexes[id], 'i');
-		});
+		}), {});
 
-		log('setting bugdb noneSelected/allSelected...');
-
-		const num_selected = size(conf.selected_app_ids);
-		db.noneSelected = (num_selected === 0);
-
-		// since allSelected is slow to eval, make it lazy
-		defineLazyProperty(db, 'allSelected', () => {
-			const num_selected_lazy = size(conf.selected_app_ids);
-			return (!!num_selected_lazy && every(db.apps, (app, app_id) => conf.selected_app_ids.hasOwnProperty(app_id)));
-		});
-
-		log('processed bugdb...');
-
-		if (!fromMemory) {
-			// if there is an older bugs object in storage update newAppIds and apply block-by-default
-			const old_bugs = conf.bugs;
-			if (old_bugs && old_bugs.hasOwnProperty('version') && (old_bugs.version !== bugs.version)) {
-				const new_app_ids = BugDb.updateNewAppIds(bugs.apps, old_bugs.apps);
-				if (new_app_ids.length > 0) {
-					log(`${new_app_ids.length} new trackers have been added`);
-					const { selected_app_ids } = conf;
-					const newBlockedTrackers = BugDb._scanForNewTrackersToBlock(bugs, old_bugs, new_app_ids, selected_app_ids);
-
-					if (newBlockedTrackers.length > 0) {
-						newBlockedTrackers.forEach((app_id) => {
-							selected_app_ids[app_id] = 1;
-						});
-
-						// (triggers a persist)
-						conf.selected_app_ids = selected_app_ids;
-					}
-					db.JUST_UPDATED_WITH_NEW_TRACKERS = true;
-				}
-			}
-
-			conf.bugs = bugs;
-		}
-
-		db.categories = BugDb._buildCategories(db);
-
-		this.db = db;
-
-		if (!skip_cache_flush) {
-			flushChromeMemoryCache();
-		}
-
-		// return true for _loadList() callback
-		return true;
-	}
-
-	/**
-	 * When the extension learns of new trackers, it has to decide whether
-	 * to block them. The logic is to look at existing trackers in the same
-	 * category: if most of them were blocked, then it makes sense to also
-	 * block the new one (majority voting). As a tie-breaker, block it iff
-	 * the whole category is blocked by default.
-	 *
-	 * @private
-	 *
-	 * @param  {Object} new_apps          trackers in the new database
-	 * @param  {Object} old_apps          trackers in the original database
-	 * @param  {array}  new_app_ids       all new trackers
-	 * @param  {array}  selected_app_ids  all new trackers
-	 * @return {array}  new trackers that should be blocked
-	 */
-	static _scanForNewTrackersToBlock(bugs, old_bugs, new_app_ids, selected_app_ids) {
-		const majority = {};
-		globals.CATEGORIES_BLOCKED_BY_DEFAULT.forEach((category) => {
-			// Bias in favor of blocking if the category will be
-			// blocked by default in a new installtion. This is
-			// mostly relevant if new categories were introduced.
-			majority[category] = 0.5;
-		});
-		Object.entries(old_bugs.apps).forEach(([app_id, { cat: category }]) => {
-			const vote = selected_app_ids[app_id] ? 1 : -1;
-			majority[category] = (majority[category] || 0) + vote;
-		});
+		this.db.bugs = engine.metadata.getPatterns().reduce((all, pattern) => ({
```

Comment:
The number of patterns isn't small, right? Be careful with the reduce-patterns. I think, it results in O(n²) complexity.

Reply:
fixed